catalog/app/components/Preview/loaders/Notebook.js

```diff
@@ -16,3 +22,72 @@ export const Loader = function NotebookLoader({ handle, children }) {
   )
   return children(utils.useErrorHandling(processed, { handle, retry: data.fetch }))
 }
+
+const IFRAME_SANDBOX_ATTRIBUTES = 'allow-scripts allow-same-origin'
+
+function waitForIframe(src) {
+  let resolved = false
+  const timeout = 30000
+
+  return new Promise((resolve, reject) => {
+    const handleError = (error) => {
+      document.body.removeChild(link)
+      reject(error)
+    }
+
+    const handleSuccess = () => {
+      resolved = true
+      document.body.removeChild(link)
+      resolve(src)
+    }
+
+    const timerId = setTimeout(() => {
+      if (resolved) return
+      handleError(new Error('Page is loading too long'))
+    }, timeout)
+
+    const link = document.createElement('iframe')
+    link.addEventListener('load', () => {
+      clearTimeout(timerId)
+      handleSuccess()
+    })
+    link.src = src
+    link.style.display = 'none'
+    link.sandbox = IFRAME_SANDBOX_ATTRIBUTES
+
+    document.body.appendChild(link)
+
+    const iframeDocument = link.contentWindow || link.contentDocuent
+    if (iframeDocument) {
+      iframeDocument.addEventListener('error', handleError)
+    }
+  })
+}
+
+async function loadVoila({ endpoint, sign, handle }) {
+  const base = `${endpoint}/voila/voila/render`
+  const url = encodeURIComponent(sign(handle))
+  const src = `${base}/?url=${url}`
+
+  // Preload iframe, then insert cached iframe
+  await waitForIframe(src)
+  return PreviewData.IFrame({ src, sandbox: IFRAME_SANDBOX_ATTRIBUTES })
+}
+
+function VoilaLoader({ handle, children }) {
+  const sign = AWS.Signer.useS3Signer()
+  const endpoint = Config.use().registryUrl
+  const data = Data.use(loadVoila, { endpoint, sign, handle })
```

Comment:
I tried to be consistent with your code. I took loaders/Pdf as a reference.

I'll move URL generation to the body of `VoilaLoader` as you suggesting. I think I shouldn't make things more complicated with hooks, because this code is a matter of change